From bc6c4174a583074198323a4385d9d2575a90a7ee Mon Sep 17 00:00:00 2001
From: Huang Chaofan <chaofan.huang@nxp.com>
Date: Fri, 26 Oct 2018 14:59:06 +0800
Subject: [PATCH 4935/5242] MLK-20093 VPU: A simple workaroud for vpu decoder
 to handle the case that some frame is not for
 diaplay

commit  7f3671ddd7e7464dcb0306c9fdd68164048959b5 from
https://source.codeaurora.org/external/imx/linux-imx.git

If the buffer is not for display, it will be released directly by
firmware but not return the event VID_API_EVENT_FRAME_BUFF_RDY for
display, for this case, just add the buffer to the queue list again

Signed-off-by: Huang Chaofan <chaofan.huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mxc/vpu-decoder-b0/vpu_b0.c |    7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/mxc/vpu-decoder-b0/vpu_b0.c b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
index 01a3e26..057acae 100755
--- a/drivers/mxc/vpu-decoder-b0/vpu_b0.c
+++ b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
@@ -1886,8 +1886,11 @@ static void vpu_api_event_handler(struct vpu_ctx *ctx, u_int32 uStrIdx, u_int32
 		if (fsrel->eType == MEDIAIP_FRAME_REQ) {
 			p_data_req = &This->vb2_reqs[fsrel->uFSIdx];
 
-			if (ctx->wait_rst_done != true && p_data_req->status != FRAME_READY)
-				vpu_dbg(LVL_ERR, "error: normal release need to set status to FRAME_RELEASE but previous status %s is not FRAME_READY\n", bufstat[p_data_req->status]);
+			if (ctx->wait_rst_done != true && p_data_req->status != FRAME_READY) {
+				vpu_dbg(LVL_WARN, "warning: normal release and previous status %s, frame not for display, queue the buffer to list again\n",
+						bufstat[p_data_req->status]);
+				list_add_tail(&p_data_req->list, &This->drv_q);
+			}
 			p_data_req->status = FRAME_RELEASE;
 		}
 		vpu_dbg(LVL_INFO, "VID_API_EVENT_REL_FRAME_BUFF uFSIdx=%d, eType=%d, size=%ld\n", fsrel->uFSIdx, fsrel->eType, sizeof(MEDIA_PLAYER_FSREL));
-- 
1.7.9.5

